File: daily/2025/3333_Find_the_Original_Typed_String_II.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    int possibleStringCount(string word, int k) {
        int mod = 1e9 + 7;
        word += '.';
        vector<int> contingous_len;
        int l = 1;
        for (int i = 1; i < word.length(); i++){
            if (word[i] == word[i-1]) l++;
            else{
                contingous_len.push_back(l);
                l = 1;
            }
        }
        long long ans = 1;
        for (int i : contingous_len){
            ans = (ans * i) % mod;
        }
        int n = contingous_len.size();
        if (n >= k) return ans;
        // for (int i : contingous_len) cout << i << " ";
        // cout << endl << endl;
        vector<int> presum(k, 0);
        presum[0] = 1;
        for (int i = 0; i < n; i++){
            int curr_len = contingous_len[i];
            vector<int> curr(k, 0);
            for (int j = 1; j < k; j++){
                // (j - curr_len) ~ (j - 1) = presum[j-1] - presum[j-curr_len-1]
                curr[j] = presum[j-1];
                if (j - curr_len - 1 >= 0) curr[j] -= presum[j - curr_len - 1];
                if (curr[j] < 0) curr[j] += mod;
            }
            presum[0] = curr[0];
            for (int j = 1; j < k; j++) presum[j] = (presum[j-1] + curr[j]) % mod;
        }
        int smaller_than_k_sum = 0;
        for (int i : presum) smaller_than_k_sum = (smaller_than_k_sum + i) % mod;
        ans -= smaller_than_k_sum;
        if (ans < 0) ans += mod;
        return ans;
    }
};

int main(){
    string word = "aabbccdd";
    int k = 7;
    // string word = "aabbccdd";
    // int k = 8;
    // string word = "aaabbb";
    // int k = 3;
    // string word = "wwwwwwwbbbbssssssssvvoooooooqqqqqqqqqvvvvvooooooocccccppppkkkkkkjnddddddbbb";
    // int k = 50;
    // string word = "bbzoggrccukkwwzqdrrhhhkvbbvjwggvvccsdwwiicumoocklyynyyhhcwjssppmmaacceehhtoowkooccekkddppxppzzyyvvll";
    // int k = 80;
    Solution S;
    cout << S.possibleStringCount(word, k) << endl;
    return 0;
}

File: daily_python/3333_Find_the_Original_Typed_String_II.py
import os
import sys
from typing import List

class Solution:
    def possibleStringCount(self, word: str, k: int) -> int:
        freq = []
        word += "."
        n = len(word)
        l = 1
        mod = 10 ** 9 + 7
        for i in range(1, n):
            if word[i] == word[i-1]:
                l += 1
            else:
                freq.append(l)
                l = 1
        ans = 1
        for f in freq:
            ans = ans * f % mod
        if len(freq) >= k:
            return ans
        presum = [1] * k
        for f in freq:
            curr = [0] * k
            for i in range(1, k):
                curr[i] = presum[i-1]
                if i - 1 - f >= 0:
                    curr[i] -= presum[i - 1 - f]
                    if curr[i] < 0:
                        curr[i] += mod
            presum[0] = curr[0]
            for i in range(1, k):
                presum[i] = presum[i-1] + curr[i] % mod
        ans = (ans - presum[k - 1] + mod) % mod
        return ans
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            word = lines[0].split('=')[1].strip()[1:-2]
            k = int(lines[1].split('=')[1].strip()[:-1])
            dataset.append((word, k))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for word, k in dataset:
        results.append(solution.possibleStringCount(word, k))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')